core/src/test/java/org/infinispan/partitionhandling/PessimisticTxPartitionHandlingReleaseLockTest.java

```diff
@@ -0,0 +1,96 @@
+import org.infinispan.commands.remote.CacheRpcCommand;
+import org.infinispan.commands.tx.PrepareCommand;
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.conflict.MergePolicy;
+import org.infinispan.distribution.MagicKey;
+import org.infinispan.remoting.inboundhandler.AbstractDelegatingHandler;
+import org.infinispan.remoting.inboundhandler.DeliverOrder;
+import org.infinispan.remoting.inboundhandler.PerCacheInboundInvocationHandler;
+import org.infinispan.remoting.inboundhandler.Reply;
+import org.infinispan.test.MultipleCacheManagersTest;
+import org.infinispan.test.TestDataSCI;
+import org.infinispan.test.TestingUtil;
+import org.infinispan.test.fwk.TransportFlags;
+import org.infinispan.transaction.LockingMode;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+import org.testng.AssertJUnit;
+import org.testng.annotations.Test;
+
+/**
+ * Reproducer for ISPN-12757
+ *
+ * @author Pedro Ruivo
+ * @since 12.1
+ */
+@Test(groups = "functional", testName = "partitionhandling.PessimisticTxPartitionHandlingReleaseLockTest")
+public class PessimisticTxPartitionHandlingReleaseLockTest extends MultipleCacheManagersTest {
+
+   private static final Log log = LogFactory.getLog(PessimisticTxPartitionHandlingReleaseLockTest.class);
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      ConfigurationBuilder builder = getDefaultClusteredCacheConfig(CacheMode.DIST_SYNC, true);
+      builder.transaction().lockingMode(LockingMode.PESSIMISTIC).useSynchronization(true);
+      builder.clustering().partitionHandling().mergePolicy(MergePolicy.NONE).whenSplit(PartitionHandling.DENY_READ_WRITES);
+      builder.clustering().remoteTimeout(Integer.MAX_VALUE); // we don't want timeouts
+      createClusteredCaches(5, TestDataSCI.INSTANCE, builder, new TransportFlags().withFD(true).withMerge(true));
+   }
+
+   public void testLockReleased() throws Exception {
+      final AdvancedCache<MagicKey, String> cache0 = this.<MagicKey, String>cache(0).getAdvancedCache();
+      final TransactionManager tm = cache0.getTransactionManager();
+      final ControlledInboundHandler handler1 = TestingUtil.wrapInboundInvocationHandler(cache(1), ControlledInboundHandler::new);
+      final MagicKey key = new MagicKey(cache0, cache(1));
+
+      Future<Void> f = fork(() -> {
+         tm.begin();
+         cache0.lock(key);
+         AssertJUnit.assertNull(cache0.get(key));
+         cache0.put(key, key.toString());
+         tm.commit();
+      });
+
+      AssertJUnit.assertTrue(handler1.receivedLatch.await(30, TimeUnit.SECONDS));
+
+      //isolate backup owner
+      TestingUtil.getDiscardForCache(manager(1)).setDiscardAll(true);
+
+      //when the new view is installed, RPC request is completed with org.infinispan.remoting.responses.CacheNotFoundResponse
+      //which triggers the partition handling
```

Comment:
Should partition handling be involved at all, if `cache0` is in a majority partition?

Also, what's blocking `cache0` from sending the commit command before the main thread installs the DISCARD protocol on manager1?

IMO you need more steps here to clarify what you expect to happen. E.g. wait for a particular cache topology in `cache0`, and verify that the transaction is stored in `PartitionHandlingManagerImpl.partialTransactions`, if that's what you mean with the comment.

Reply:
> Should partition handling be involved at all, if cache0 is in a majority partition?

to be on the safer side, yes. The `MultiTargetRequest` is updated in the JGroups view change event but the availability mode is only updated after the coordinator computes the partition. So, I don't think we can make a decision. Just give the transaction to the PartitionHandlingManager and let it decide when the stable topology is installed.

> Also, what's blocking cache0 from sending the commit command before the main thread installs the DISCARD protocol on manager1?

there is no commit and the prepare is blocked by `ControlledInboundHandler`

> IMO you need more steps here to clarify what you expect to happen.

hmmm ok. I didn't want to add complexity to the test.

Reply:
> to be on the safer side, yes. The MultiTargetRequest is updated in the JGroups view change event but the availability mode is only updated after the coordinator computes the partition. So, I don't think we can make a decision. Just give the transaction to the PartitionHandlingManager and let it decide when the stable topology is installed.

Does this mean `StateTransferInterceptor` no longer retries commit/1PC prepare commands?
I see the code is still there, but maybe it only runs with partition handling disabled.

> hmmm ok. I didn't want to add complexity to the test.

From my point of view it is *simplifying* the reading of the test :)

BTW, I find the new comment a bit confusing as well

```
//after the view change, the transaction is handled by the PartitionHandlingManager which finishes the commit
```

After reading the whole method a couple of times I understood that the transaction is passed to PartitionHandlingManager and the commit call returns, but the first time I thought it meant "wait until PartitionHandlingManager retries the commit and the transaction is completed"

Reply:
> Does this mean `StateTransferInterceptor` no longer retries commit/1PC prepare commands?
> I see the code is still there, but maybe it only runs with partition handling disabled.

No, Partition handling only handles a single case: a crash from a node involved in the transaction. All other cases (leaves/join/crashes from other nodes) are retried.

> After reading the whole method a couple of times I understood that the transaction is passed to PartitionHandlingManager and the commit call returns, but the first time I thought it meant "wait until PartitionHandlingManager retries the commit and the transaction is completed"

Sorry about that :)

Reply:
> No, Partition handling only handles a single case: a crash from a node involved in the transaction. All other cases (leaves/join/crashes from other nodes) are retried.

I see, I mistook your "Just give the transaction to the PartitionHandlingManager" comment to mean that PHM will always take care of completing the transaction after a `CacheNotFoundResponse`.